### ISSUE TITLE
crashes node and chrome browser when parsing PNG-compressed icon

### ISSUE THREAD

**Issue by gkovacs** [egy186/icojs#18]
Steps to reproduce:

Download http://news.ycombinator.com/favicon.ico and save it as favicon.ico

Run the sample code on node:

```
const fs = require('fs');
const ICO = require('icojs');
 
const arrayBuffer = new Uint8Array(fs.readFileSync('favicon.ico')).buffer;
ICO.parse(arrayBuffer).then(images => {
  // do something 
});
```

Stacktrace:

```
<--- Last few GCs --->

[5572:0x103000000]    23518 ms: Mark-sweep 1214.6 (1252.9) -> 1199.6 (1252.9) MB, 897.1 / 0.0 ms  allocation failure GC in old space requested
[5572:0x103000000]    24539 ms: Mark-sweep 1199.6 (1252.9) -> 1199.6 (1252.9) MB, 1020.5 / 0.0 ms  allocation failure GC in old space requested
[5572:0x103000000]    25587 ms: Mark-sweep 1199.6 (1252.9) -> 1199.6 (1220.9) MB, 1048.1 / 0.0 ms  last resort
[5572:0x103000000]    26627 ms: Mark-sweep 1199.6 (1220.9) -> 1199.6 (1220.9) MB, 1040.0 / 0.0 ms  last resort


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x34078dc26729 <JS Object>
    2: extractOne(aka extractOne) [/Users/geza/Dropbox/habitlab-macbookair/node_modules/icojs/src/extract-one.js:~20] [pc=0xbbe146ea0e5](this=0x38ce95e02311 <undefined>,buffer=0x3574204c7dd1 <an ArrayBuffer with map 0x1e5dc84030c9>,index=0)
    3: /* anonymous */(aka /* anonymous */) [/Users/geza/Dropbox/habitlab-macbookair/node_modules/icojs/src/ico.js:35] [pc=0xbbe146e3940](this=0x38ce95e0231...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 4: v8::internal::Factory::NewUninitializedFixedArray(int) [/usr/local/bin/node]
 5: v8::internal::(anonymous namespace)::ElementsAccessorBase<v8::internal::(anonymous namespace)::FastPackedObjectElementsAccessor, v8::internal::(anonymous namespace)::ElementsKindTraits<(v8::internal::ElementsKind)2> >::GrowCapacity(v8::internal::Handle<v8::internal::JSObject>, unsigned int) [/usr/local/bin/node]
 6: v8::internal::Runtime_GrowArrayElements(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/bin/node]
 7: 0xbbe144043a7
 8: 0xbbe1469a8cc
fish: 'node foo.js' terminated by signal SIGABRT (Abort)
```

When run from the background script of a chrome extension, it also crashes chrome during the icojs.parse() step (just crashes and doesn't give any stack trace)

```
fetch('http://news.ycombinator.com/favicon.ico').then(x => x.arrayBuffer()).then(x =>result = icojs.parse(x))
```

This works fine for other favicons (ie, the favicon from slashdot and several other major sites work fine), seems to just be specific to this particular favicon

If relevant, I'm on node 7.6.0 on macOS Sierra 10.12.3

**Comment by gkovacs** [comment]
also I just built the version from master, and the same issue occurs there too. appears to be running out of memory on the javascript heap.

```
geza@gezamac ~/D/icojs> node ./foo.js

<--- Last few GCs --->

[6255:0x102801600]      356 ms: Scavenge 8.2 (13.5) -> 6.9 (18.0) MB, 1.8 / 0.0 ms  allocation failure
[6255:0x102801600]      439 ms: Scavenge 12.1 (19.5) -> 10.5 (20.0) MB, 2.4 / 0.0 ms  allocation failure
[6255:0x102801600]      529 ms: Scavenge 13.2 (20.5) -> 11.4 (22.0) MB, 4.3 / 0.0 ms  allocation failure
[6255:0x102801600]      653 ms: Scavenge 15.8 (23.5) -> 13.4 (32.0) MB, 3.2 / 0.0 ms  allocation failure


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x1cb13d426729 <JS Object>
    1: InnerArrayFill(aka InnerArrayFill) [native array.js:~1037] [pc=0x242cc4875edc](this=0x2da0fd682311 <undefined>,K=0,at=0x2da0fd682311 <undefined>,au=167772262,q=0x166defdf749 <JS Array[167772262]>,r=167772262)
    2: fill [native array.js:1] [pc=0x242cc4875b89](this=0x166defdf749 <JS Array[167772262]>,K=0,at=0x2da0fd682311 <undefined>,au=0x2da0fd682311 <undefined>)
    3: arguments adapto...

FATAL ERROR: invalid table size Allocation failed - JavaScript heap out of memory
 1: node::Abort() [/usr/local/bin/node]
 2: node::FatalException(v8::Isolate*, v8::Local<v8::Value>, v8::Local<v8::Message>) [/usr/local/bin/node]
 3: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [/usr/local/bin/node]
 4: v8::internal::HashTable<v8::internal::SeededNumberDictionary, v8::internal::SeededNumberDictionaryShape, unsigned int>::EnsureCapacity(v8::internal::Handle<v8::internal::SeededNumberDictionary>, int, unsigned int, v8::internal::PretenureFlag) [/usr/local/bin/node]
 5: v8::internal::SeededNumberDictionary::AddNumberEntry(v8::internal::Handle<v8::internal::SeededNumberDictionary>, unsigned int, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyDetails, bool) [/usr/local/bin/node]
 6: v8::internal::(anonymous namespace)::DictionaryElementsAccessor::AddImpl(v8::internal::Handle<v8::internal::JSObject>, unsigned int, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, unsigned int) [/usr/local/bin/node]
 7: v8::internal::JSObject::AddDataElement(v8::internal::Handle<v8::internal::JSObject>, unsigned int, v8::internal::Handle<v8::internal::Object>, v8::internal::PropertyAttributes, v8::internal::Object::ShouldThrow) [/usr/local/bin/node]
 8: v8::internal::Runtime::SetObjectProperty(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::LanguageMode) [/usr/local/bin/node]
 9: v8::internal::Runtime_KeyedStoreIC_Slow(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/bin/node]
10: 0x242cc46043a7
fish: 'node ./foo.js' terminated by signal SIGABRT (Abort)
```

**Comment by egy186** [comment]
Thanks for reporting. I can reproduce this issue.

The crash occurs when ICO includes PNG as ICONIMAGE, which is not supported in v0.5.0.
My plan is...
- `v0.6.0`: reject with error when PNG is included (rather than crash)
- `v0.7.0`: support parsing ICO which includes PNG

**Comment by egy186** [comment]
Fixed in version 0.7.0.
